### ISSUE TITLE
Ignore inspircd chanhistory messages on login

### ISSUE THREAD

**Issue by docwhat** [cinchrb/cinch#193]
When connecting to an inspircd server with the [chanhistory](https://wiki.inspircd.org/Modules/2.0/chanhistory) plugin turned on, then the history for the channel is _replayed_.

Cinch should ignore these old messages on initial connect.  Instead what it does is respond to all of them, which is very annoying.

I suspect you can just wait until the timestamp actually matches the current time on the messages, but I'm not an IRC protocol guru.

Ciao!

**Comment by docwhat** [comment]
I suppose a silly hack would be to tell cinch not to reply to anything for *X* seconds on connect.  Not sure how to do that since I'm using Cinch via Lita's IRC adapter.

**Comment by dominikh** [comment]
Can you provide me with a public inspircd server and channel that has chanhistory turned on?

**Comment by docwhat** [comment]
hmm...not easily, @dominikh 

I'm a dab hand at Ruby... is there a way I could dump information for you instead? Like a simple test bot that dumps out everything? I could sanitize the output and post it here.

Worse case scenario, I could set up one just for you someplace, I guess...

**Comment by dominikh** [comment]
Cinch by default logs all IRC traffic to stderr. If Lita changes that, you'll have to consult with Lita. Or just run a vanilla Cinch bot for this. Download http://sprunge.us/YQIZ set the correct address/channel, run the file and send me the log output. If there's channel or server passwords involved, you will want to remove these from the logs. Please redact as little as possible to avoid changing important parts.

**Comment by docwhat** [comment]
@dominikh I sent you an email (from your profile page) with the log output because I don't want it shared. I hope that is okay.

**Comment by dominikh** [comment]
I'm not sure why I expected more from inspircd, but their chanhistory module is awful for bots. It doesn't include any markers indicating the start or end of the replay, it doesn't include timestamps (IRC itself doesn't have timestamps), and it doesn't tell you how many messages it is going to replay to you, either. ("Up to 20 lines" is as useful as "Some lines").

The only way to filter these replayed messages would indeed be to ignore all messages the first X milliseconds after joining the channel. But that's not only fragile and vulnerable to lag, it would also risk losing new messages sent in that period. As such, it's not something I want to add to Cinch.

In theory, the "ignore messages for X milliseconds" bit can be implemented in the individual plugins. Listen to the join event for the bot, note the current time, then for each message check its time against the join event's time (+ some careful code to avoid race conditions). However, this isn't a viable solution when using 3rd party plugins.

You could ease the pain by running the bot behind a proxy, such as ZNC. The proxy would stay connected at all times, and the bot would connect to the proxy. Restarting the bot would thus not send the channel history again. Of course you'd still run into the issue once the bot joins a new channel.

I know this won't help you, but in the end this should be a bug report/feature request on the inspircd project. They should be implementing [Message tags](https://github.com/ircv3/ircv3-specifications/blob/master/specification/message-tags-3.2.md) and optionally [server-time](https://github.com/ircv3/ircv3-specifications/blob/master/extensions/server-time-3.2.md) and then the replay could be ignored easily.

**Comment by docwhat** [comment]
How about ignoring messages until the MODE is sent. I noticed it is sent last. 

I'll see about filing a bug for inspircd.

**Comment by dominikh** [comment]
Re "MODE": For all we know, that's only true by chance. Also, doing that would tie it directly to unrelated logic in Cinch. The server is sending us the MODE because we request it. And between us requesting it, and the server responding, we might receive new channel messages.

**Comment by docwhat** [comment]
It looks like if a user marked as a bot then channel history isn't replayed: https://github.com/inspircd/inspircd/pull/616

I'll try that out on Monday. I'll also see about adding your suggestions on their issues tracker, too.

**Comment by docwhat** [comment]
Sorry for the long reply... That change isn't in the current stable inspircd, version 2.0.18.  So the bot flag doesn't work. :-(

I'm looking to backport [that patch](https://github.com/inspircd/inspircd/commit/71b9af776ef53c6902c020cf729c5277b824e0f9) to my inspircd.

Meanwhile, I'm trying to figure out hubot isn't spammed by the history... (or if it is, how I didn't notice before).

**Comment by docwhat** [comment]
Okay, so I took the meat of that patch (see below) and while it *sort of* works, it also doesn't.

While any channels that cinch joins don't replay, there is one channel (like a `#general`) that everyone is forced to join... that force-joined group is always replaying its contents.

I think that is a bug in inspircd, though.

I figured out how hubot wasn't having this problem. I had a "<connection>" section in the configuration that forced the `+B` on connections via specific port/password that eliminates the race condition. So I have a work around and am going to close this.

I think this is fair to say it's an inspircd problem, not a cinch problem.

Thanks for your help, though...
